src/cost.lagda.md

```diff
@@ -0,0 +1,97 @@
+# Cost
+
+Here I will try to define a cost function and then I will prove the cost of a parallel reduction of a tree using the cost approach.
+
+
+
+# Imports
+Import the needed libraries
+
+```agda
+module cost
+ where
+
+open import Data.Unit.Base
+open import Data.Product
+open import Data.Nat.Base
+open import Agda.Builtin.Equality
+open import Relation.Binary.PropositionalEquality
+open ≡-Reasoning
+
+variable
+  A B C D : Set
+```
+# Cartesian Data Structure
+
+Define the closed cartesian category data structure and operations over the structure.
+```agda
+data _⟹_ : Set → Set → Set₁ where
+  mid : A ⟹ A
+  _m∘_ : (f : B ⟹ C) → (g : A ⟹ B) → (A ⟹ C)
+  m! : A ⟹ ⊤
+  mexl : (A × B) ⟹ A
+  mexr :  (A × B) ⟹ B
+  _m▵_ : (left : A ⟹ C) → (right : A ⟹ D) → (A ⟹ (C × D))
+  madd : ( ℕ × ℕ ) ⟹ ℕ
+
+evaluate : (f : A ⟹ B) → (x : A) → B
+evaluate mid x = x
+evaluate (f m∘ g) x = evaluate f (evaluate g x)
+evaluate m! x = tt
+evaluate mexl ( a , b ) =  a
+evaluate mexr (a , b) = b
+evaluate (f m▵ g) x = (evaluate f x , evaluate g x)
+evaluate madd (a , b) = a + b
+
+parallelCost : (A ⟹ B) → A → ℕ
+parallelCost mid a = 0
+parallelCost (f m∘ g) a = parallelCost f (evaluate g a) + parallelCost g a
+parallelCost m! a = 0
+parallelCost mexl a = 0
+parallelCost mexr a = 0
+parallelCost (f m▵ g) a = (parallelCost f a) ⊔ (parallelCost g a)
+parallelCost (madd) x = 1
+```
+
```

Comment:
What should the ideal parallel time ("depth") be for `mexl m∘ (f m▵ g)` be?

Reply:
> I see what you are getting at. Under my previous defintion, it would be `(cost f) ⊔ (cost g)`, when acutally it should be at most `cost f`, beacuse g is discarded.

That’s right!

> This makes me think that finding the cost of an expression is undecidable, and so I cannot use a function to calculate it, but rather should use a proposition.

Before you give up on computing cost (depth and/or work), try to refine your style. Reflect on what prevents you from arranging for the automatically computed cost of `mexl m∘ (f m▵ g)` being equal to the cost of `f`?

Reply:
> This makes me think that finding the cost of an expression is undecidable, and so I cannot use a function to calculate it, but rather should use a proposition.

More fundamental than the question of whether cost is computable is whether it is definable. What would a proposition that a computation `f` costs `x` *mean*?